checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java

```diff
@@ -201,6 +209,86 @@ private void checkOwningOverrides(
     }
   }
 
+  /**
+   * If a {@code @MustCallAlias} annotation appears in a method declaration, it must appear as an
+   * annotation on both the return type, and a parameter type.
+   *
+   * <p>The return type is checked if it is annotated with {@code @PolyMustCall} because the Must
+   * Call Checker treats {@code @MustCallAlias} as an alias of {@code @PolyMustCall}.
+   *
+   * @param tree the method declaration.
+   * @param mcAtf the MustCallAnnotatedTypeFactory.
+   */
+  private void checkMustCallAliasAnnotationForMethod(
+      MethodTree tree, MustCallAnnotatedTypeFactory mcAtf) {
+
+    boolean isMustCallAliasAnnoOnParameter = isMustCallAliasAnnoPresentInParams(tree, mcAtf);
+
+    if (TreeUtils.isVoidReturn(tree) && isMustCallAliasAnnoOnParameter) {
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param");
+      return;
+    }
+
+    AnnotatedTypeMirror returnType = mcAtf.getMethodReturnType(tree);
+    boolean isMustCallAliasAnnoOnReturnType = returnType.hasPrimaryAnnotation(PolyMustCall.class);
+
+    if (isMustCallAliasAnnoOnParameter != isMustCallAliasAnnoOnReturnType) {
+      checker.reportWarning(tree, "mustcallalias.method.return.and.param");
+    }
+  }
+
+  /**
+   * Given a constructor, a {@code @MustCallAlias} must appear in both the list of parameters and as
+   * an annotation on the constructor itself, if it is to appear at all.
+   *
+   * <p>That is, a {@code @MustCallAlias} annotation must appear on both the constructor and its
+   * parameter list, or not at all.
+   *
+   * @param tree the constructor.
+   * @param mcAtf a MustCallAnnotatedTypeFactory.
+   */
+  private void checkMustCallAliasAnnotationForConstructor(
+      MethodTree tree, MustCallAnnotatedTypeFactory mcAtf) {
+    ExecutableElement constructorDecl = TreeUtils.elementFromDeclaration(tree);
+    boolean isMustCallAliasAnnoOnConstructor =
+        constructorDecl != null
+            && constructorDecl.getAnnotationMirrors().stream()
+                .anyMatch(anno -> atypeFactory.areSameByClass(anno, MustCallAlias.class));
```

Comment:
Rather than this implementation, I think you could instead reuse `ResourceLeakAnnotatedTypeFactory#hasMustCallAlias(Element)`, whose implementation is:
```
  /*package-private*/ boolean hasMustCallAlias(Element elt) {
    if (noResourceAliases) {
      return false;
    }
    MustCallAnnotatedTypeFactory mustCallAnnotatedTypeFactory =
        getTypeFactoryOfSubchecker(MustCallChecker.class);
    return mustCallAnnotatedTypeFactory.getDeclAnnotationNoAliases(elt, MustCallAlias.class)
        != null;
  }
```

I don't think that's directly accessible from `MustCallVisitor`, but I think it would be okay to move the implementation into the Must Call Checker and then have the existing method call a delegate.

Reply:
I'm a bit confused here: the code you commented on is implemented within the `ResourceLeakVisitor`, and not the `MustCallChecker`:

> I think it would be okay to move the implementation into the Must Call Checker and then have the existing method call a delegate.

Do you mean I should move the code I've written (i.e., to check that the `@MustCallAlias` annotation occurs in pairs) into the Must Call Checker? Or did you mean for me to move the implementation of the `ResourceLeakAnnotatedTypeFactory#hasMustCallAlias(Element)` into the Must Call Checker?

Reply:
> I'm a bit confused here: the code you commented on is implemented within the ResourceLeakVisitor, and not the MustCallChecker:

This was my mistake - I thought I was reading `MustCallVisitor`. You can disregard that part of the comment.

> I think `ResourceLeakAnnotatedTypeFactory#hasMustCallAlias(Element)` is accessible from the `ResourceLeakVisitor`

It is, and I think you should call it instead of writing new code that does the same job. Sorry for the confusion!